File: 算法/Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        m_stack = []  # make list as a stack
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '[' or s[i] == '{':  # push open brackets
                m_stack.append(s[i])
            elif len(m_stack) == 0:  # quantity of open brackets  < close brackets
                return False
            else:
                # open brackets not be closed in the correct order.
                if (s[i] == ')' and m_stack[-1] != '(') or (s[i] == '}' and m_stack[-1] != '{') or (s[i] == ']' and m_stack[-1] != '['):
                    return False
                else:
                    m_stack.pop()
        if len(m_stack) != 0:  # uantity of open brackets  > close brackets
            return False
        return True

File: 算法/C++/20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> m_stack;
        for (int i = 0; i < s.size(); i++) {
            switch (s[i]) {
            case '(':
            case '{':
            case '[': m_stack.push(s[i]); break; // push Open brackets
            case ')':
                if (m_stack.empty() || m_stack.top() != '(') // there isn't an Open brackets matching  close brackets
                    return false;
                else
                    m_stack.pop();
                break;
            case '}':
                if (m_stack.empty() || m_stack.top() != '{')
                    return false;
                else
                    m_stack.pop();
                break;
            case ']':
                if (m_stack.empty() || m_stack.top() != '[')
                    return false;
                else
                    m_stack.pop();
                break;
            }
        }
        if (!m_stack.empty()) // left extra open brackets
            return false;
        return true;
    }
};